File: git_ops.py
#!/usr/bin/env python3
"""Git operations for agentic coding assistant worktree management."""

import subprocess
import os
import re
from pathlib import Path
from typing import Optional

from utils import C_GOOD, C_INFO, C_NORMAL, cprint


class GitError(Exception):
    """Git operation failed."""
    pass


def run_git(*args, cwd=None, check=True, capture_output=True):
    """Run a git command and return the result."""
    cmd = ['git'] + list(args)
    result = subprocess.run(
        cmd,
        cwd=cwd,
        capture_output=capture_output,
        text=True,
        check=False
    )
    if check and result.returncode != 0:
        raise GitError(f"Git command failed: {' '.join(cmd)}\n{result.stderr}")
    return result


def is_git_repo(path='.'):
    """Check if the given path is inside a git repository."""
    result = run_git('rev-parse', '--git-dir', cwd=path, check=False)
    return result.returncode == 0


def init_git_repo(path='.'):
    """Initialize a new git repository."""
    run_git('init', cwd=path)
    # Create an initial commit to have a main branch
    readme_path = Path(path) / 'README.md'
    if not readme_path.exists():
        readme_path.write_text('# Project\n\nInitialized by maca.\n')
    run_git('add', 'README.md', cwd=path)
    run_git('commit', '-m', 'Initial commit', cwd=path)


def get_repo_root(path='.'):
    """Get the root directory of the git repository."""
    result = run_git('rev-parse', '--show-toplevel', cwd=path)
    return Path(result.stdout.strip())


def get_current_branch(cwd='.'):
    """Get the name of the current branch."""
    result = run_git('rev-parse', '--abbrev-ref', 'HEAD', cwd=cwd)
    return result.stdout.strip()


def get_head_commit(cwd='.'):
    """Get the current HEAD commit hash."""
    result = run_git('rev-parse', 'HEAD', cwd=cwd)
    return result.stdout.strip()


# def get_commits_between(old_commit, new_commit, cwd='.'):
#     """
#     Get list of commits between old_commit and new_commit.

#     Returns list of dicts with 'hash' and 'message' (first line only).
#     """
#     # Format: <hash> <first line of message>
#     result = run_git('log', '--format=%H %s', f'{old_commit}..{new_commit}', cwd=cwd, check=False)

#     if result.returncode != 0 or not result.stdout.strip():
#         return []

#     commits = []
#     for line in result.stdout.strip().split('\n'):
#         if line:
#             parts = line.split(' ', 1)
#             if len(parts) == 2:
#                 commits.append({
#                     'hash': parts[0][:8],  # Short hash
#                     'message': parts[1]
#                 })

#     return commits


# def get_changed_files_between(old_commit, new_commit, cwd='.'):
#     """
#     Get list of files changed between old_commit and new_commit.

#     Returns list of file paths.
#     """
#     result = run_git('diff', '--name-only', old_commit, new_commit, cwd=cwd, check=False)

#     if result.returncode != 0 or not result.stdout.strip():
#         return []

#     return [line.strip() for line in result.stdout.strip().split('\n') if line.strip()]


def find_next_session_id(repo_root):
    """Find the next available session ID by checking .maca directory."""
    maca_dir = Path(repo_root) / '.maca'
    maca_dir.mkdir(exist_ok=True)

    # Find all existing session directories
    existing = []
    for item in maca_dir.iterdir():
        if item.is_dir() and item.name.isdigit():
            existing.append(int(item.name))

    return max(existing, default=0) + 1


def create_session_worktree(repo_root, session_id):
    """Create a new branch and worktree for the session."""
    branch_name = f'maca/{session_id}'
    
    maca_dir = Path(repo_root) / '.maca'
    maca_dir.mkdir(parents=True, exist_ok=True)

    worktree_path = maca_dir / str(session_id)

    # Get current branch to branch from
    current_branch = get_current_branch(cwd=repo_root)

    # Clean up state worktrees and branches
    run_git('worktree', 'prune', cwd=repo_root)

    # Create new branch
    run_git('branch', '-f', branch_name, current_branch, cwd=repo_root)

    # Create worktree
    run_git('worktree', 'add', str(worktree_path), branch_name, cwd=repo_root)

    # Create .scratch directory for temporary analysis files
    scratch_dir = worktree_path / '.scratch'
    scratch_dir.mkdir(exist_ok=True)

    cprint(
        C_GOOD, f'Session {session_id} created',
        C_NORMAL, ' (branch: ', C_INFO, branch_name,
        C_NORMAL, ', worktree: ', C_INFO, str(worktree_path.relative_to(repo_root)), C_NORMAL, ')',
    )

    return worktree_path, branch_name


def commit_changes(worktree_path, message):
    """Commit all changes in the worktree with the given message, excluding .scratch and .maca."""
    # Add all changes (including untracked files), but exclude .scratch and .maca
    run_git('add', '-A', ':!.scratch', ':!.maca', cwd=worktree_path)

    # Commit
    return run_git('commit', '-m', message, cwd=worktree_path, check=False).returncode == 0


def generate_descriptive_branch_name(commit_message):
    """
    Generate a descriptive branch name from commit message.

    Returns a branch name (without maca/ prefix).
    """
    # Extract first line of commit message
    first_line = commit_message.split('\n')[0].strip()

    # Remove common prefixes
    for prefix in ['Add', 'Update', 'Fix', 'Remove', 'Refactor', 'Implement']:
        if first_line.startswith(prefix + ' '):
            first_line = first_line[len(prefix) + 1:]
            break

    # Convert to branch name format (lowercase, hyphenated, max 40 chars)
    name = first_line.lower()
    name = re.sub(r'[^a-z0-9\s-]', '', name)  # Remove special chars
    name = re.sub(r'\s+', '-', name)  # Replace spaces with hyphens
    name = re.sub(r'-+', '-', name)  # Collapse multiple hyphens
    name = name.strip('-')  # Remove leading/trailing hyphens
    name = name[:40]  # Limit length
    name = name.rstrip('-')  # Remove trailing hyphen if truncated

    # Ensure we have something
    if not name:
        name = 'changes'

    return name


def merge_to_main(root_path, worktree_path, org_branch_name, commit_message):
    """Merge the session branch into main using squash + rebase + ff strategy."""
    root_branch = get_current_branch(cwd=root_path)
    worktree_branch = get_current_branch(cwd=worktree_path)

    # Generate descriptive branch name for preserving history
    descriptive_branch = org_branch_name + '-' + generate_descriptive_branch_name(commit_message)
    run_git('checkout', '-b', descriptive_branch, cwd=worktree_path, check=False) # May err if we rerun after rebase conflict

    # Get the merge base
    base_commit = run_git('merge-base', root_branch, worktree_branch, cwd=worktree_path).stdout.strip()

    # Append preservation note to commit message
    enhanced_message = commit_message.rstrip() + f'\n\nThe original chain of MACA commits is kept in the {descriptive_branch} branch.'

    # Soft reset to base
    run_git('reset', '--soft', base_commit, cwd=worktree_path)

    # Stage all changes (excluding .scratch and .maca)
    run_git('add', '-A', ':!.scratch', ':!.maca', cwd=worktree_path)

    # Commit everything as one commit with enhanced message
    result = run_git('commit', '-m', enhanced_message, cwd=worktree_path, check=False)
    if result.returncode != 0:
        # If commit fails, it might be because there are no changes (edge case)
        # Just continue without erroring
        if 'nothing to commit' in result.stdout + result.stderr:
            return None
        else:
            # Real error, re-raise
            raise GitError(f"Git command failed: git commit\nstdout: {result.stdout}\nstderr: {result.stderr}")

    # Try to rebase the session branch onto main
    result = run_git('rebase', root_branch, cwd=worktree_path, check=False)

    if result.returncode != 0:
        return result.stdout

    # Fast-forward merge the rebased descriptive branch
    run_git('merge', '--ff-only', descriptive_branch, cwd=root_path)


def cleanup_session(repo_root, worktree_path, branch_name):
    """Clean up the worktree and branch after merge."""
    # Remove worktree
    run_git('worktree', 'remove', str(worktree_path), cwd=repo_root, check=False)

    # Delete branch
    run_git('branch', '-D', branch_name, cwd=repo_root, check=False)


def reset_worktree_to_main(repo_root, worktree_path, branch_name):
    """Reset the worktree and branch to match main (for reuse)."""
    main_branch = get_current_branch(cwd=repo_root)

    # In the worktree, reset hard to main
    run_git('fetch', 'origin', main_branch, cwd=worktree_path, check=False)
    run_git('reset', '--hard', main_branch, cwd=worktree_path)

    # Clean untracked files
    run_git('clean', '-fd', cwd=worktree_path)

File: src/maca/git_ops.nim
## Git operations for agentic coding assistant worktree management

import std/[osproc, os, strutils, re, sequtils]

type
  GitError* = object of CatchableError

  CommitInfo* = object
    hash*: string
    message*: string

proc runGit*(args: varargs[string], cwd = "", check = true): tuple[output: string, exitCode: int] =
  ## Run a git command and return the result
  let cmd = "git " & args.join(" ")
  let workDir = if cwd == "": getCurrentDir() else: cwd

  let (output, exitCode) = execCmdEx(cmd, workingDir = workDir)

  if check and exitCode != 0:
    raise newException(GitError, "Git command failed: " & cmd & "\n" & output)

  return (output.strip(), exitCode)

proc isGitRepo*(path = "."): bool =
  ## Check if the given path is inside a git repository
  let (_, exitCode) = runGit(["rev-parse", "--git-dir"], cwd = path, check = false)
  return exitCode == 0

proc initGitRepo*(path = ".") =
  ## Initialize a new git repository
  discard runGit(["init"], cwd = path)

  # Create an initial commit to have a main branch
  let readmePath = path / "README.md"
  if not fileExists(readmePath):
    writeFile(readmePath, "# Project\n\nInitialized by maca.\n")

  discard runGit(["add", "README.md"], cwd = path)
  discard runGit(["commit", "-m", "\"Initial commit\""], cwd = path)

proc getRepoRoot*(path = "."): string =
  ## Get the root directory of the git repository
  let (output, _) = runGit(["rev-parse", "--show-toplevel"], cwd = path)
  return output

proc getCurrentBranch*(cwd = "."): string =
  ## Get the name of the current branch
  let (output, _) = runGit(["rev-parse", "--abbrev-ref", "HEAD"], cwd = cwd)
  return output

proc getHeadCommit*(cwd = "."): string =
  ## Get the current HEAD commit hash
  let (output, _) = runGit(["rev-parse", "HEAD"], cwd = cwd)
  return output

proc getCommitsBetween*(oldCommit, newCommit: string, cwd = "."): seq[CommitInfo] =
  ## Get list of commits between oldCommit and newCommit
  ##
  ## Returns seq of CommitInfo with 'hash' and 'message' (first line only)

  let (output, exitCode) = runGit(["log", "--format=%H %s", oldCommit & ".." & newCommit], cwd = cwd, check = false)

  if exitCode != 0 or output == "":
    return @[]

  result = @[]
  for line in output.splitLines():
    if line != "":
      let parts = line.split(' ', 1)
      if parts.len == 2:
        result.add(CommitInfo(
          hash: parts[0][0..7],  # Short hash
          message: parts[1]
        ))

proc getChangedFilesBetween*(oldCommit, newCommit: string, cwd = "."): seq[string] =
  ## Get list of files changed between oldCommit and newCommit

  let (output, exitCode) = runGit(["diff", "--name-only", oldCommit, newCommit], cwd = cwd, check = false)

  if exitCode != 0 or output == "":
    return @[]

  return output.splitLines().filterIt(it != "")

proc findNextSessionId*(repoRoot: string): int =
  ## Find the next available session ID by checking .maca directory
  let macaDir = repoRoot / ".maca"
  createDir(macaDir)

  # Find all existing session directories
  var existing: seq[int] = @[]
  for kind, path in walkDir(macaDir):
    if kind == pcDir:
      let name = path.lastPathPart()
      try:
        existing.add(parseInt(name))
      except ValueError:
        discard

  if existing.len > 0:
    return max(existing) + 1
  else:
    return 1

proc createSessionWorktree*(repoRoot: string, sessionId: int): tuple[worktreePath: string, branchName: string] =
  ## Create a new branch and worktree for the session
  let branchName = "maca-" & $sessionId
  let sessionDir = repoRoot / ".maca" / $sessionId
  let worktreePath = sessionDir / "worktree"

  # Ensure session directory exists
  createDir(sessionDir)

  # Get current branch to branch from
  let currentBranch = getCurrentBranch(cwd = repoRoot)

  # Clean up stale worktrees and branches
  discard runGit(["worktree", "prune"], cwd = repoRoot)

  # Create new branch
  discard runGit(["branch", "-f", branchName, currentBranch], cwd = repoRoot)

  # Create worktree
  discard runGit(["worktree", "add", worktreePath, branchName], cwd = repoRoot)

  # Create .scratch directory for temporary analysis files
  let scratchDir = worktreePath / ".scratch"
  createDir(scratchDir)

  return (worktreePath, branchName)

proc commitChanges*(worktreePath: string, message: string): bool =
  ## Commit all changes in the worktree with the given message, excluding .scratch and .maca

  # Add all changes (including untracked files), but exclude .scratch and .maca
  discard runGit(["add", "-A", ":!.scratch", ":!.maca"], cwd = worktreePath)

  # Check if there are changes to commit
  let (_, exitCode) = runGit(["diff", "--cached", "--quiet"], cwd = worktreePath, check = false)
  if exitCode == 0:
    # No changes to commit
    return false

  # Commit
  discard runGit(["commit", "-m", "\"" & message.replace("\"", "\\\"") & "\""], cwd = worktreePath)
  return true

proc getDiffStats*(worktreePath: string): string =
  ## Get the diff statistics for uncommitted changes
  let (output, _) = runGit(["diff", "--numstat", "HEAD"], cwd = worktreePath, check = false)
  return output

proc generateDescriptiveBranchName*(commitMessage, repoRoot: string): string =
  ## Generate a descriptive branch name from commit message
  ##
  ## Returns a branch name that doesn't conflict with existing maca/* branches

  # Extract first line of commit message
  var firstLine = commitMessage.splitLines()[0].strip()

  # Remove common prefixes
  for prefix in ["Add", "Update", "Fix", "Remove", "Refactor", "Implement"]:
    if firstLine.startsWith(prefix & " "):
      firstLine = firstLine[prefix.len + 1..^1]
      break

  # Convert to branch name format (lowercase, hyphenated, max 40 chars)
  var name = firstLine.toLowerAscii()
  name = name.replace(re"[^a-z0-9\s-]", "")  # Remove special chars
  name = name.replace(re"\s+", "-")  # Replace spaces with hyphens
  name = name.replace(re"-+", "-")  # Collapse multiple hyphens
  name = name.strip(chars = {'-'})  # Remove leading/trailing hyphens
  if name.len > 40:
    name = name[0..39]
  name = name.strip(trailing = true, chars = {'-'})  # Remove trailing hyphen if truncated

  # Ensure we have something
  if name == "":
    name = "changes"

  # Check if maca/<name> exists, if so add suffix
  let baseName = name
  var counter = 2
  while true:
    let fullBranch = "maca/" & name
    let (_, exitCode) = runGit(["rev-parse", "--verify", fullBranch], cwd = repoRoot, check = false)
    if exitCode != 0:
      # Branch doesn't exist, we can use it
      break
    # Branch exists, try next variant
    name = baseName & "-" & $counter
    inc counter

  return name

proc mergeToMain*(repoRoot, worktreePath, branchName, commitMessage: string): tuple[success: bool, message: string] =
  ## Merge the session branch into main using squash + rebase + ff strategy
  let mainBranch = getCurrentBranch(cwd = repoRoot)

  # Switch to the session branch in the main repo
  discard runGit(["checkout", branchName], cwd = repoRoot)

  # Save the current HEAD commit hash (before squashing)
  let (originalHead, _) = runGit(["rev-parse", "HEAD"], cwd = repoRoot)

  # Get the merge base
  let (baseCommit, _) = runGit(["merge-base", mainBranch, branchName], cwd = repoRoot)

  # Generate descriptive branch name for preserving history
  let descriptiveName = generateDescriptiveBranchName(commitMessage, repoRoot)

  # Append preservation note to commit message
  var enhancedMessage = commitMessage
  if not enhancedMessage.endsWith("\n"):
    enhancedMessage &= "\n"
  enhancedMessage &= "\nThe original chain of MACA commits is kept in the maca/" & descriptiveName & " branch."

  # Soft reset to base
  discard runGit(["reset", "--soft", baseCommit], cwd = repoRoot)

  # Commit everything as one commit with enhanced message
  discard runGit(["commit", "-m", "\"" & enhancedMessage.replace("\"", "\\\"") & "\""], cwd = repoRoot, check = false)

  # Go back to main branch
  discard runGit(["checkout", mainBranch], cwd = repoRoot)

  # Try to rebase the session branch onto main
  let (_, rebaseExitCode) = runGit(["rebase", mainBranch, branchName], cwd = repoRoot, check = false)

  if rebaseExitCode != 0:
    # Rebase failed, likely due to conflicts
    # Abort the rebase
    discard runGit(["rebase", "--abort"], cwd = repoRoot, check = false)
    return (false, "Merge conflicts detected")

  # Fast-forward merge
  discard runGit(["merge", "--ff-only", branchName], cwd = repoRoot)

  # Create the descriptive branch pointing at original HEAD
  discard runGit(["branch", "maca/" & descriptiveName, originalHead], cwd = repoRoot)

  return (true, "Merged successfully")

proc cleanupSession*(repoRoot, worktreePath, branchName: string) =
  ## Clean up the worktree and branch after merge

  # Remove worktree
  discard runGit(["worktree", "remove", worktreePath], cwd = repoRoot, check = false)

  # Delete branch
  discard runGit(["branch", "-D", branchName], cwd = repoRoot, check = false)
